### ISSUE TITLE
What is daggy for?

### ISSUE THREAD

**Issue by edahlseng** [fantasyland/daggy#31]
I've seen daggy used in many different examples across the Internet, but I'm still having trouble understanding what it's for. I've read the [Why "daggy" issue](https://github.com/fantasyland/daggy/issues/4), but that seems to talk more about why daggy over another library. I understand that daggy is useful for making "sum types," which makes sense in the abstract, but I'm having trouble formalizing in a practical context. What does using this library give us over just having different objects that could be combined at a type level such as with [Flow union types](https://flow.org/en/docs/types/unions/), or even some other alternative?

**Comment by davidchambers** [comment]
It gives us types such as this:

```haskell
data Blah a b = Foo a | Bar b | Baz a b | Quux
```

Crucially, it gives us `cata` to allow us to perform case folding:

```haskell
Blah.cata :: { Foo :: a -> c, Bar :: b -> c, Baz :: a -> b -> c, Quux :: c } -> Blah -> c
```

As a bonus, `cata` asserts that all the cases are provided (totality checking).

**Comment by edahlseng** [comment]
@davidchambers thanks for the information! After some additional research of my own, I learned that sum types/tagged unions are not the exact same thing as ordinary union types, which I think contributed to some of my confusion.

Out of curiosity, why the names `daggy` and `cata`?

If you're open to it, I'd like to distill some of this information to add to the README in order to help explain to others who might stumble on this repository in the future. I can put up a PR this week!

**Comment by puffnfresh** [comment]
@edahlseng that's right, tagged unions aren't like e.g. C unions. I was gonna call the library like "taggy" or something, but that sounded like "daggy" to me, which is an Australian word for unfashionable. Doesn't really mean anything.

Cata is short for "catamorphism" which means the fold function over a data type.

**Comment by edahlseng** [comment]
Thanks @puffnfresh!

I've recently discovered how to use [Flow for disjoint unions](https://flow.org/en/docs/types/unions/#toc-disjoint-unions). I'm wondering, what are the pros and cons between that approach and daggy? My current understanding is that the real benefit of sum types comes from compile-time checks, which seems to be a pro for Flow. This is all still new to me, so I'm still learning a lot about daggy and sum types in general.

Thanks for being so helpful and responsive!

**Comment by joedski** [comment]
My favorite use of Daggy is for [representing the instantaneous state of single asynchronous request](https://medium.com/javascript-inside/slaying-a-ui-antipattern-in-react-64a3b98242c).  As a bonus, it's the easiest way I can ~~trick~~ convince coworkers to let me add it to our projects.
